Xoriant/m1

shashanknarang27 opened the pull request:
We are creating the PR for Xoriant Phase-1 /Milestone-1 . We have covered 20 ADT messages.

BoyaWu10 commented on src/Microsoft.Health.Fhir.Liquid.Converter.FunctionalTests/FunctionalTests.cs:
```diff
@@ -24,7 +24,49 @@ public static IEnumerable<object[]> GetDataForHl7v2()
         {
             var data = new List<string[]>
             {
+                new[] { @"ADT_A01", @"ADT_A01.hl7", @"ADT-A01-expected.json" },
```
The functional tests fail for all the sample data and templates. Could you please double check about it? The tests could be found in `Visual Studio 2019 - Test Explorer - Microsoft.Health.Fhir.Liquid.Converter.FunctionalTests.FunctionalTests`.

shashanknarang27 replied:
We have pushed the changes for the test case failure. There are 40 new cases which has been added and 13 which already present. It is expected that those 13 test case will fail as the template code has been changed.

We have added one condition to check null reference in FunctionalTest.cs as we are not creating "div" attribute in "Provenance" resource and while running the existing code it is giving null reference. So check this , we have added the condition.
`new List<string>
            {
                "$.entry[?(@.resource.resourceType=='Provenance')].resource.text.div",
            }.ForEach(path =>
            {
                JToken expectedObjectToken = expectedObject.SelectToken(path,false);
                if(expectedObjectToken != null) 
                {
                    expectedObject.SelectToken(path,false).Parent.Remove();
                }
                JToken actualObjectToken = actualObject.SelectToken(path,false);
                if(actualObjectToken != null)
                {
                    actualObject.SelectToken(path,false).Parent.Remove();
                }
            });`

BoyaWu10 replied:
For "Provenance" resource, I saw that we are removing the "div" generation by commenting them out [here](https://github.com/microsoft/FHIR-Converter/blob/c13e23ded1bfef1bf75496b6690b0e55b5259084/data/Templates/Hl7v2/Resource/_Provenance.liquid#L27). Are we expecting to do so? The "div" here is designed to record the template information like template version. If we do not remove it, we would not need to add this check.

BoyaWu10 replied:
If we have modified the existing templates, we can update the expected results of original test cases accordingly.

shashanknarang27 replied:
For the "Provenance" resource , as per our SME's there were no mapping for "div" so we have commented that. As you mention that "div" was designed to record the template information, now we will add it the provenance resource created from MSH segment.
Provenance resource which are created from other segments like EVN will not have this "div". 
We will remove the null check condition from FunctionTests.cs which we have added.

We will update the expected result of original test cases once get the confirmation from Gina.

shashanknarang27 replied:
@BoyaWu10 : We have added the "div" attribute to provenance and reverted the null condition check. We have updated the expected json for the 13 test case.

BoyaWu10 reviewed (commented):
In the latest codes, we are using a JSON file [CodeSystem.json](https://github.com/microsoft/FHIR-Converter/blob/xoriant/m1/data/Templates/Hl7v2/CodeSystem/CodeSystem.json) to record all the mappings for CodeSystems. The liquid templates are removed and not used anymore. Instead of adding new liquid templates for CodeSystems, could you please add the mappings to the JSON file? The details could be found [here](https://github.com/microsoft/FHIR-Converter/blob/xoriant/m1/docs/SnippetConcept.md#hl7-v2-specific-snippets).

qiwjin commented:
/azp run

qiwjin commented:
/azp run

shashanknarang27 commented on src/Microsoft.Health.Fhir.Liquid.Converter.FunctionalTests/FunctionalTests.cs:
```diff
@@ -81,45 +123,30 @@ public static IEnumerable<object[]> GetDataForCcda()
             var expectedContent = File.ReadAllText(expectedFile);
             var traceInfo = new Hl7v2TraceInfo();
-            var actualContent = hl7v2Processor.Convert(inputContent, rootTemplate, new TemplateProvider(templateDirectory, DataType.Hl7v2), traceInfo);
+            var actualContent = hl7v2Processor.Convert(inputContent, rootTemplate, new Hl7v2TemplateProvider(templateDirectory), traceInfo);
```
We have rebased the code and now using TemplateProvider in FunctionalTests.cs files.

shashanknarang27 commented:
> In the latest codes, we are using a JSON file [CodeSystem.json](https://github.com/microsoft/FHIR-Converter/blob/xoriant/m1/data/Templates/Hl7v2/CodeSystem/CodeSystem.json) to record all the mappings for CodeSystems. The liquid templates are removed and not used anymore. Instead of adding new liquid templates for CodeSystems, could you please add the mappings to the JSON file? The details could be found [here](https://github.com/microsoft/FHIR-Converter/blob/xoriant/m1/docs/SnippetConcept.md#hl7-v2-specific-snippets).

We have already added the codesystem mapping in JSON file "CodeSystem.json". The template can be use to refer what mapping exist.
All the liquid template code mapping is present in JSON file "CodeSystem.json".

BoyaWu10 commented:
> > In the latest codes, we are using a JSON file [CodeSystem.json](https://github.com/microsoft/FHIR-Converter/blob/xoriant/m1/data/Templates/Hl7v2/CodeSystem/CodeSystem.json) to record all the mappings for CodeSystems. The liquid templates are removed and not used anymore. Instead of adding new liquid templates for CodeSystems, could you please add the mappings to the JSON file? The details could be found [here](https://github.com/microsoft/FHIR-Converter/blob/xoriant/m1/docs/SnippetConcept.md#hl7-v2-specific-snippets).
> 
> We have already added the codesystem mapping in JSON file "CodeSystem.json". The template can be use to refer what mapping exist.
> All the liquid template code mapping is present in JSON file "CodeSystem.json".

Thanks! Then we can just remove all the CodeSystem liquid templates except the sample one (_SampleUniversalIDType.liquid).

BoyaWu10 commented:
Besides, some unit tests fail as well in [CI](https://microsofthealthoss.visualstudio.com/FhirConverter/_build/results?buildId=14868&view=results). Could you please check them as well?

shashanknarang27 commented:
> > > In the latest codes, we are using a JSON file [CodeSystem.json](https://github.com/microsoft/FHIR-Converter/blob/xoriant/m1/data/Templates/Hl7v2/CodeSystem/CodeSystem.json) to record all the mappings for CodeSystems. The liquid templates are removed and not used anymore. Instead of adding new liquid templates for CodeSystems, could you please add the mappings to the JSON file? The details could be found [here](https://github.com/microsoft/FHIR-Converter/blob/xoriant/m1/docs/SnippetConcept.md#hl7-v2-specific-snippets).
> > 
> > 
> > We have already added the codesystem mapping in JSON file "CodeSystem.json". The template can be use to refer what mapping exist.
> > All the liquid template code mapping is present in JSON file "CodeSystem.json".
> 
> Thanks! Then we can just remove all the duplicate CodeSystem liquid templates except the sample one (_SampleUniversalIDType.liquid).

We have removed the liquid template files and put the sample one (_SampleUniversalIDType.liquid).

shashanknarang27 commented:
> Besides, some unit tests fail as well in [CI](https://microsofthealthoss.visualstudio.com/FhirConverter/_build/results?buildId=14868&view=results). Could you please check them as well?

We have updated the Unit tests and Functional Test and all the test are passing now. 
Build got failed during pipeline execution on generating notice file. We are not sure on this why it has failed.
Let us know if any thing need to be changed from our side on this.

BoyaWu10 commented on src/Microsoft.Health.Fhir.Liquid.Converter.UnitTests/Models/Hl7v2/Hl7v2TraceInfoTests.cs:
```diff
@@ -62,15 +62,15 @@ public void GivenHl7v2Data_WhenCreate_CorrectHl7v2TraceInfoShouldBeReturned()
             var templateProvider = new TemplateProvider(TestConstants.Hl7v2TemplateDirectory, DataType.Hl7v2);
             _ = processor.Convert(content, "ADT_A01", templateProvider, traceInfo);
-            Assert.Single(traceInfo.UnusedSegments);
+           //Assert.Single(traceInfo.UnusedSegments);
```
Nit: Can we update the expected number of unused segments here instead of commenting it out?

shashanknarang27 replied:
We have uncommented and added the check for expected no of unused Segments

qiwjin commented:
/azp run

BoyaWu10 commented:
> > Besides, some unit tests fail as well in [CI](https://microsofthealthoss.visualstudio.com/FhirConverter/_build/results?buildId=14868&view=results). Could you please check them as well?
> 
> We have updated the Unit tests and Functional Test and all the test are passing now.
> Build got failed during pipeline execution on generating notice file. We are not sure on this why it has failed.
> Let us know if any thing need to be changed from our side on this.

Thanks! We have updated the CI pipeline and the step of notice file generation is skipped now.

BoyaWu10 commented:
Can we add the new cases in [rule-based tests](https://github.com/microsoft/FHIR-Converter/blob/ffa85e3c817a28bd20ec6bc87eefe66aec70dcdb/src/Microsoft.Health.Fhir.Liquid.Converter.FunctionalTests/RuleBasedTests.cs#L45) as well?

qiwjin commented:
/azp run

BoyaWu10 commented:
> Can we add the new cases in [rule-based tests](https://github.com/microsoft/FHIR-Converter/blob/ffa85e3c817a28bd20ec6bc87eefe66aec70dcdb/src/Microsoft.Health.Fhir.Liquid.Converter.FunctionalTests/RuleBasedTests.cs#L45) as well?

Hi @shashanknarang27, could you please update the codes to the latest xoriant/m1? We have removed some tests that take too long, which lead to time out in CI.

qiwjin commented on src/Microsoft.Health.Fhir.TemplateManagement.UnitTests/Providers/TemplateCollectionProviderTests.cs:
```diff
@@ -85,8 +85,8 @@ public static IEnumerable<object[]> GetImageInfoForArtifactWithTag()
         {
-            yield return new object[] { "microsofthealth/fhirconverter:default", 759 };
-            yield return new object[] { "microsofthealth/hl7v2templates:default", 759 };
+            yield return new object[] { "microsofthealth/fhirconverter:default", 834 };
```
I feel we can refine the hard code numbers in future. @sowu880

shashanknarang27 commented:
> Can we add the new cases in [rule-based tests](https://github.com/microsoft/FHIR-Converter/blob/ffa85e3c817a28bd20ec6bc87eefe66aec70dcdb/src/Microsoft.Health.Fhir.Liquid.Converter.FunctionalTests/RuleBasedTests.cs#L45) as well?

We have added new RuleBase test cases for 20 ADT messages

shashanknarang27 commented:
@BoyaWu10 : Should we rebase the code to latest xoriant/m1 ?  We have seen that there are few changes been pushed to xoriant/m1

BoyaWu10 commented:
> @BoyaWu10 : Should we rebase the code to latest xoriant/m1 ? We have seen that there are few changes been pushed to xoriant/m1

Both would be OK! We have pushed some commits to xoriant/m1 to reduce CI running time. It has already taken effect in current PR. :smile:
